File: Graph/rotten_oranges.cc
// Link: https://leetcode.com/problems/rotting-oranges/

// Approach: The general idea is to apply BFS and mark all the healthy oranges near rotten orange as rotten. However, you first have to fill the queue with all the rotten
// oranges (I didn't did at first) because oranges rot in parallel and not the series. So apply BFS simultaneously on rotten oranges. Now, calculate the level of BFS and
// return that. Because it is the time to rot all the oranges (Or max level to traverse all nodes). I have used the variables "c1" and "c2" to calculate levels since I am
// not using recursion. Variable "c1" stores number of parent nodes and "c2" stores number of child nodes. We decrement "c1" as parent node traverses and if it becomes zero,
// we change level (variable named "time" here).

class Solution
{
    private:
        queue<pair<int, int>> node_queue;
        
        void print_grid(vector<vector<int>>& grid)
        {
            int m = grid.size(), n = grid[0].size();
            
            for(int i=0;i<m;i++)
            {
                for(int j=0;j<n;j++)
                {
                    cout<<grid[i][j]<<", ";
                }
                cout<<"\n";
            }
            cout<<"---------------------------------\n";
        }
    
        int bfs(int& one_counter, vector<vector<int>>& grid)
        {
            pair<int, int> p;
            int x = -1, y = -1, m = grid.size(), n = grid[0].size(), time = 0;
            int c1 = node_queue.size(), c2 = 0;
            // node_queue.push(make_pair(i, j));
            
            while(node_queue.size()>0)
            {
                p = node_queue.front();
                x = p.first;
                y = p.second;
                node_queue.pop();
                c1--;
                
                if(x>0)
                {
                    if(grid[x-1][y] == 1)
                    {
                        node_queue.push(make_pair(x-1, y));
                        grid[x-1][y] = -1;
                        one_counter--;
                        c2++;
                    }
                }
                
                if(x<(m-1))
                {
                    if(grid[x+1][y] == 1)
                    {
                        node_queue.push(make_pair(x+1, y));
                        grid[x+1][y] = -1;
                        one_counter--;
                        c2++;
                    }
                }
                
                if(y>0)
                {
                    if(grid[x][y-1] == 1)
                    {
                        node_queue.push(make_pair(x, y-1));
                        grid[x][y-1] = -1;
                        one_counter--;
                        c2++;
                    }
                }
                
                if(y<(n-1))
                {
                    if(grid[x][y+1] == 1)
                    {
                        node_queue.push(make_pair(x, y+1));
                        grid[x][y+1] = -1;
                        one_counter--;
                        c2++;
                    }
                }
                
                if(c1 == 0)
                {
                    c1 = c2;
                    c2 = 0;
                    time++;
                    
                    // print_grid(grid);
                    // cout<<time<<"\n---------\n";
                }
            }
            return --time;
        }
    
    public:
        int orangesRotting(vector<vector<int>>& grid)
        {
            int m = grid.size();
            
            if(m == 0)
            {
                return 0;
            }
            
            int n = grid[0].size(), time = 0, one_counter = 0;
            
            for(int i=0;i<m;i++)
            {
                for(int j=0;j<n;j++)
                {
                    if(grid[i][j] == 2)
                    {
                        node_queue.push(make_pair(i, j));
                    }
                    
                    if(grid[i][j] == 1)
                    {
                        one_counter++;
                    }
                }
            }
            
            if(one_counter == 0 && node_queue.size() == 0)
            {
                return 0;
            }
            else
            {
                time = bfs(one_counter, grid);
                
                if(one_counter != 0)
                {
                    return -1;
                }
            }
            return time;
        }
};

File: Graph/rotten_oranges.py
# Link: https://practice.geeksforgeeks.org/problems/rotten-oranges2536/1

# Approach: First explore all rotten oranges and put their co-ordinates into queue. Also, keep count of good oranges during that. Now, do BFS on all the rotten oranges and make fresh oranges rot during
# BFS. Do BFS by counting llevels and at each level end, increment the time. Do this until no rotten oranges is left in the queue. Also, simultaneously decrement fresh orange count as you make them rot.
# Finally if no fresh orange is left, then return the time, else return -1.

from collections import deque

def rotOranges(mat,r,c):
    rot_time, orange_counter, queue = -1, 0, deque()
    neighbor_idx = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    for i in range(r):
        for j in range(c):
            if mat[i][j] == 1:
                orange_counter += 1 # Counts Fresh Orange
            elif mat[i][j] == 2:
                queue.append((i, j))
    
    while queue:
        qlen = len(queue)
        
        while qlen:
            cell = queue.popleft()
            
            for i, j in neighbor_idx:
                ni, nj = cell[0]+i, cell[1]+j
                
                if (0 <= ni < r) and (0 <= nj < c) and (mat[ni][nj] == 1):
                    orange_counter, mat[ni][nj] = orange_counter - 1, 2
                    queue.append((ni, nj))
            
            qlen -= 1
        
        rot_time += 1
    
    return (rot_time if not orange_counter else -1)